design-proposals/klip-19-materialize-views.md

```diff
@@ -0,0 +1,223 @@
+* Re-purpose `CREATE STREAM` to create a mutable stream backed by a new or existing topic.
+    * `INSERT`s into the stream will be supported.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of any new topic. 
+* Replace `CREATE STREAM AS SELECT` with `CREATE MATERIALIZED STREAM VIEW AS SELECT`.
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of the sink topic.   
+* Either `CREATE STREAM AS SELECT` will be removed, or maintained as creating a mutable derived stream. 
+    * `INSERT`s into the stream will be supported.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of the sink topic.
+    
+For those worried about the excessive typing required for `MATERIALIZED STREAM VIEW`, we could add a shorthand `CREATE MSV` etc.
+
+## Design
+
+Design is mostly covered above by public API changes. Changes to KSQL to implement this KLIP should be straight forward.
+
+All new terminology will be supported in interactive-mode. Headless mode will support all new statements, but will 
+remove support for `CREATE TABLE` statements, as these are only of use when combined with `INSERT VALUES` statements, 
+which, are not yet supported in headless mode.
+
+The proposed new syntax for tables is Ansi SQL compliant, though obviously the `WITH` syntax part if not.
+(This incompatibility could be fixed by defining defaults for formats and using source names for topics,
+ but this is out of scope for this KLIP).
+
+The new `VIEW` variants will support the same `WITH` properties that their non-`VIEW` variants do today, 
+unless otherwise stated in the [Public APIS section](#public_apis) above. 
+
+The CLI or Server will be updated to detect any removed statements and inform the user they are no longer supported,
+and what to do instead, linking to more documentation. This can be done through a general pattern that can 
+be used for other removed statement types.
+
+## Test plan
+
+Proposed changes in this KLIP are mostly just syntax changes, with the exception of fixing semantics for `INSERT VALUES`.
+Hence, updating existing test cases will cover most of the work.  Test will be added to cover fixes to `INSERT VALUES`.
+
+Not system / muckrake tests will be added.
+
+## Documentation Updates
+
+Existing documentation will be overhauled to take new syntax into account, with special care taken to call out:
+ * `VIEW`s being readonly
+ * The undefined behaviour if `CREATE TABLE` is used with a topic that is updated by another system
+ * Proper table update semantics for `INSERT VALUES` for non-`VIEW` sources.
+
+Existing demos will need updating. 
+
+Updating existing blogs will be left up to the blog author. 
+
+# Compatibility Implications
+
+The KLIP proposes removing support for `CREATE TABLE AS SELECT`, and potentially `CREATE STREAM AS SELECT`.
```

Comment:
Why removing instead of deprecation?

Reply:
Because we can make breaking changes at the moment as we're not guaranteeing any backwards compatibility. Of course, we could also just deprecate and remove when we move to v1.0, but that carries more tech debt in the meantime.

Reply:
Maybe I am more conservative being used to the Kafka process. But KSQL is around for a few years already and even if "ksqlDB" has a lower version, it has a big user base already. I would always prefer to not break backward compatibility if possible. Just my 2 cents.